server/libs/infra/src/search/typesense.repository.ts

```diff
@@ -0,0 +1,322 @@
+import {
+  ISearchRepository,
+  SearchCollection,
+  SearchCollectionIndexStatus,
+  SearchFilter,
+  SearchResult,
+} from '@app/domain';
+import { Injectable, Logger } from '@nestjs/common';
+import _, { Dictionary } from 'lodash';
+import { Client } from 'typesense';
+import { CollectionCreateSchema } from 'typesense/lib/Typesense/Collections';
+import { DocumentSchema, SearchResponse } from 'typesense/lib/Typesense/Documents';
+import { AlbumEntity, AssetEntity } from '../db';
+import { albumSchema } from './schemas/album.schema';
+import { assetSchema } from './schemas/asset.schema';
+
+interface GeoAssetEntity extends AssetEntity {
+  geo?: [number, number];
+}
+
+function removeNil<T extends Dictionary<any>>(item: T): Partial<T> {
+  _.forOwn(item, (value, key) => {
+    if (_.isNil(value) || (_.isObject(value) && !_.isDate(value) && _.isEmpty(removeNil(value)))) {
+      delete item[key];
+    }
+  });
+
+  return item;
+}
+
+const schemaMap: Record<SearchCollection, CollectionCreateSchema> = {
+  [SearchCollection.ASSETS]: assetSchema,
+  [SearchCollection.ALBUMS]: albumSchema,
+};
+
+const schemas = Object.entries(schemaMap) as [SearchCollection, CollectionCreateSchema][];
+
+interface SearchUpdateQueue<T = any> {
+  upsert: T[];
+  delete: string[];
+}
+
+@Injectable()
+export class TypesenseRepository implements ISearchRepository {
+  private logger = new Logger(TypesenseRepository.name);
+  private queue: Record<SearchCollection, SearchUpdateQueue> = {
+    [SearchCollection.ASSETS]: {
+      upsert: [],
+      delete: [],
+    },
+    [SearchCollection.ALBUMS]: {
+      upsert: [],
+      delete: [],
+    },
+  };
+
+  private _client: Client | null = null;
+  private get client(): Client {
+    if (!this._client) {
+      throw new Error('Typesense client not available (no apiKey was provided)');
+    }
```

Comment:
Is there a way to just fail in the constructor so we cannot have the scenario of `_client` being null?

Reply:
That would require the service not being initialized when typesense is disabled. That means prevent dependent injection conditionally, which I'm not sure how to do, so this is what I came up with instead.